### ISSUE TITLE
Postgres database option

### ISSUE THREAD

**Issue by brettinternet** [binwiederhier/ntfy#1114]
:bulb: **Postgres!**
I use ntfy in a kubernetes environment and would like the option of high availability. I also prefer to use a central DB connection over a volume with a sqlite DB.

:computer: **ntfy server**

**Comment by fredrikscode** [comment]
First result I found while googling "ntfy postgres". I'd love the option of going with postgres or any other type of database server as sqlite bricks over nfs and using volumes is just a very crappy way to deal with storage when running clusters.

**Comment by Ittihadyya** [comment]
Seconding this, I would prefer being able to slot it into already existing infrastructure than having it run its own SQLite database.

**Comment by giangduong** [comment]
+1

**Comment by sadorowo** [comment]
I already use Postgres for various services. It would be nice to connect ntfy to it, instead of having separate databases for no reason.

**Comment by gavinmcfall** [comment]
Also throwing my weight into this one. would love to make this more highly available

**Comment by PixNyb** [comment]
Not being able to use an external database doesn't work in my environment and as such keeps me from enjoying the functionalities that ntfy.sh does very well. Can't wait to see this added

**Comment by binwiederhier** [comment]
Due to popular demand, I may look into this. FYI I did a relatively straight forward PoC a while back: https://github.com/binwiederhier/ntfy/pull/1095

**Comment by hugosxm** [comment]
I use gotify now, with the iOS app that is not working :( I really prefer NTFY and I need pgsql support for that because I run k8s too like many people i guess... So big thank you @binwiederhier ! In the meantime I think I will switch back to ntfy with a local volume !

**Comment by binwiederhier** [comment]
I have started to implement this here: https://github.com/binwiederhier/ntfy/pull/1611/changes

The design is described in the docs:

```
By default, ntfy uses SQLite for all database-backed stores. As an alternative, you can configure ntfy to use PostgreSQL
by setting the `database-url` option to a PostgreSQL connection string:

database-url: "postgres://user:pass@host:5432/ntfy"

When `database-url` is set, ntfy will use PostgreSQL for the web push subscription store, the user manager and the
message cache instead of SQLite. The `web-push-file` option is not required in this case.

You can also set this via the environment variable `NTFY_DATABASE_URL` or the command line flag `--database-url`.
```

Does this sound reasonable?

**Comment by k8ieone** [comment]
Yess, looking forward to deploying this! ❤️

**Comment by RivenSkaye** [comment]
Due to how the current `*-file` keys are set up, what would happen if both those keys _and_ a postgres URI are specified?
This _could be_ very useful as a migration path for existing users who would prefer PostgreSQL.

For example:
- Read in data from sqlite files, [`INSERT ... ON CONFLICT DO UPDATE`](https://www.postgresql.org/docs/18/sql-insert.html) on the PostgreSQL DB and warn the user that they should _probably_ be removing the keys from their config.
- To prevent service restart edge cases, delete the sqlite files.
- If the config isn't updated, skip the migration code and warn the user about pre-migration remnants instead.

**Comment by binwiederhier** [comment]
The code is code complete and working well. I have to do a thorough round of regression testing of the entire product before releasing it, but if anyone wants to test it, go wild: https://github.com/binwiederhier/ntfy/pull/1619

It was all done by AI, I have manually reviewed this branch over and over and over again, and I have compared before/after. I also had Cursor created this nice doc for me with a comparison, which I re-did 5-10 times: [comparison-2026-03-01.html](https://github.com/user-attachments/files/25717103/comparison-2026-03-01.html)

--

> Due to how the current *-file keys are set up, what would happen if both those keys and a postgres URI are specified?
This could be very useful as a migration path for existing users who would prefer PostgreSQL.

The ntfy server will not allow both to be set. It's either `database-url` or any of the `*-file` options. 

**Migration:**   
I cannot offer a permanent migration path, that would be insane to have to maintain. I had Cursor write a one-off migration tool (pgimport) in the tools folder of that PR. I tested this with the prod ntfy databases and it seems to work. I have not tested that thoroughly though. 

**Test plan:**

- Deploy code (unchanged settings, using sqlite) to ntfy staging, using the prod database
- Test _all the things_ on ntfy staging
- Stop ntfy server on ntfy staging
- Perform the pgimport, update configs to `database-url`
- Start ntfy server on ntfy staging
- Test _all the things_ on ntfy staging
- Perform a load test on ntfy staging, with load similar to the prod ntfy server

**Release plan:**

- Assuming all this passes, tag new release, announce it
- Wait 3-5 days, do not upgrade prod ntfy (let community test sqlite changes + postgres changes)
- Upgrade prod ntfy server (using old sqlite config)
- Wait ~7 days (let community test postgres changes)
- Stop ntfy server on ntfy prod
- Perform the pgimport, update configs to `database-url`
- Start ntfy server on ntfy prod
- Monitor

**Comment by binwiederhier** [comment]
Everyone, please test ntfy 2.18.0 and the Postgres support: https://github.com/binwiederhier/ntfy/releases/tag/v2.18.0

You can learn how to configure it here: https://docs.ntfy.sh/config/#postgresql-experimental

TLDR: `database-url: "postgres://user:pass@host:5432/ntfy"`

**Comment by k8ieone** [comment]
Switched my instance today without any issues.

I didn't do any imports, fresh start. I didn't have anything persistent on my instance that I wouldn't mind losing.

We'll see how it holds up, my instance is mainly used as a UnifiedPush distributor.

Thank you for your work on this @binwiederhier ❤️

**Comment by binwiederhier** [comment]
Thanks for the feedback @k8ieone

**Comment by cmuench** [comment]
I was able to import two sqlite dbs into my postgtres server with the pgimport tool. Everything works well.
Thanks @k8ieone

**Comment by FAUSheppy** [comment]
Worked like a charm so far, will test the cli/acls & stuff later this week. Thanks for this great feature!

**Comment by binwiederhier** [comment]
Thank you for testing folks. Keep the feedback coming! I'll likely switch ntfy.sh next weekend unless I hear anything that will make me stop it.

**Comment by FAUSheppy** [comment]
Only minor complaint about the error handling:

1. When putting in an invalid database URL, it seemingly silently defaults to a localhost connection and this combined with...
2. ..the error handling [here](https://github.com/binwiederhier/ntfy/blob/bcd07115c24faf7dc6074a876950070fa40bfeef/db/pg/pg.go#L64), is a bit awkward because it doesn't say, that the **Database** Ping failed.

This left me confused about the actual problem.

Idk what [this check](https://github.com/binwiederhier/ntfy/blob/bcd07115c24faf7dc6074a876950070fa40bfeef/db/pg/pg.go#L28) does exactly, but with this URL it did break (meaning it didn't fail where it should):

    postgresql+psycopg2://notification-management:PW@8.8.8.8/notification-management?sslmode=require

(the URL is wrong, I'm just yapping about the unclear error)

**Comment by binwiederhier** [comment]
@FAUSheppy That's great feedback. I hate bad error messages. I'll improve this.

**Comment by binwiederhier** [comment]
@FAUSheppy Done: https://github.com/binwiederhier/ntfy/commit/997e20fa3f70920097b4deb376b7f2f38b6c6dc3

**Comment by kyuuk** [comment]
just migrated to postgres using the `pgimport` tool (users.db), everything is good, tested some notification and nothing to report except that it works 😄

**Comment by hugosxm** [comment]
Sorry for the delay, tested everything today, works like a charm but did not test the migration as I took the opportunity to start from a fresh instance and brand new token / secrets. Thanks a lot for that change ! 
Is this possible to cache attachements to postgres too ? It would make ntfy fully stateless ! 

Also can we run multiple replicas of ntfy with postgres ?

**Comment by binwiederhier** [comment]
> Is this possible to cache attachements to postgres too ? 

My plan was S3 bucket/adapter, but I guess pg could be an idea too. 

> Also can we run multiple replicas of ntfy with postgres ?

I merged in DB replica support yesterday (not released yet):

```
database-url: "postgres://user:pass@primary:5432/ntfy?sslmode=require&sslrootcert=/etc/ntfy/db-ca-cert.pem&pool_max_conns=30"
database-replica-urls:
  - "postgres://user:pass@replica1:5432/ntfy?sslmode=require&sslrootcert=/etc/ntfy/db-ca-cert.pem&pool_max_conns=30"
  - "postgres://user:pass@replica2:5432/ntfy?sslmode=require&sslrootcert=/etc/ntfy/db-ca-cert.pem&pool_max_conns=30"
```

You can NOT yet run multiple instances of ntfy yet. There is in-memory state that prevents that. I will add that once I have attachments and replicas fully rolled out.

The postgres push was just step one into making it clusterable or at least more horizontally scalable.

**Comment by k8ieone** [comment]
omg, yesss! You're building towards dream HA!

**Comment by hugosxm** [comment]
> > Is this possible to cache attachements to postgres too ? 
> 
> My plan was S3 bucket/adapter, but I guess pg could be an idea too. 
> 
> > Also can we run multiple replicas of ntfy with postgres ?
> 
> I merged in DB replica support yesterday (not released yet):
> 
> ```
> database-url: "postgres://user:pass@primary:5432/ntfy?sslmode=require&sslrootcert=/etc/ntfy/db-ca-cert.pem&pool_max_conns=30"
> database-replica-urls:
>   - "postgres://user:pass@replica1:5432/ntfy?sslmode=require&sslrootcert=/etc/ntfy/db-ca-cert.pem&pool_max_conns=30"
>   - "postgres://user:pass@replica2:5432/ntfy?sslmode=require&sslrootcert=/etc/ntfy/db-ca-cert.pem&pool_max_conns=30"
> ```
> 
> You can NOT yet run multiple instances of ntfy yet. There is in-memory state that prevents that. I will add that once I have attachments and replicas fully rolled out.
> 
> The postgres push was just step one into making it clusterable or at least more horizontally scalable.
> 

Thanks for the clarification, postgres seems easier as a user if you already have the DB. S3 is nice too but this is an additional component to run if you not already have it. 

I would be glad to test HA once it is ready ;)

**Comment by dvjn** [comment]
Hey @binwiederhier 

Love seeing the strides being made toward making ntfy HA-ready!

Just wanted to check, is there an issue or milestone where I can track HA support progress? I can spot individual efforts like PostgreSQL and S3 support, but wasn't sure if there's somewhere to follow the bigger picture. Totally understand if it's not tracked that way yet, just thought it might be a nice addition!

**Comment by binwiederhier** [comment]
It's only tracked in my head. There is only one piece missing: making the shared state and message forwarding between multiple ntfy instances work. It should not be difficult, but it's a little bit involved.

**Comment by noon92** [comment]
Hi! I'm another HA booster. I was just about to hack together something that barely works and only for my use case, but it sounds like you might be close to real HA?

**Comment by binwiederhier** [comment]
I'll have to do it eventually as ntfy.sh grows, though there is a lot of headroom now with postgres, the access cache and s3 for storage. 

And with Microsoft, Google and such being down a lot, people wouldn't care if ntfy was down for an hour. I think I may have a better uptime than GitHub haha.

**Comment by noon92** [comment]
Thanks for the quick response! 

In my specific case, I use NTFY to manage a homelab, but those servers also host a webserver for a veterinary ambulance service, including new patient intake forms. When a client orders an ambulance, they fill out the form and I get notifications about the sick dog or cat via NTFY (and email, but NTFY is the default). The webserver is cloned automatically to a remote which serves it if my main is down, and I get notifications of the main being down. Of course, if the main is fully down (and not just the webserver) then NTFY is down too. I'd love to have it fail over to my secondary. That way patient intakes will always arrive, as will notifications of the main having issues. 


NTFY is amazing. I assume you're extremely proud. If you're not, you should be! Thank you - your tool literally is helping to save lives.

**Comment by binwiederhier** [comment]
I love these types of stories. That definitely encourages me to make HA work ;-)
